File: src/main/ruby/g0001_0100/s0094_binary_tree_inorder_traversal/solution.rb
# #Easy #Top_100_Liked_Questions #Top_Interview_Questions #Depth_First_Search #Tree #Binary_Tree
# #Stack #Data_Structure_I_Day_10_Tree #Udemy_Tree_Stack_Queue #Big_O_Time_O(n)_Space_O(n)
# #2023_11_22_Time_56_ms_(84.00%)_Space_211.1_MB_(13.60%)

require_relative '../../com_github_leetcode/tree_node'

# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer[]}
def inorder_traversal(root)
  # puts "node = #{root&.val}"

  return [] if root.nil?

  nodes_visited = []

  unless root.left.nil?
    left_nodes = inorder_traversal(root.left)
    nodes_visited.concat(left_nodes)
  end

  nodes_visited << root.val

  unless root.right.nil?
    right_nodes = inorder_traversal(root.right)
    nodes_visited.concat(right_nodes)
  end

  nodes_visited
end

File: src/main/scala/g0001_0100/s0094_binary_tree_inorder_traversal/Solution.scala
package g0001_0100.s0094_binary_tree_inorder_traversal

// #Easy #Top_100_Liked_Questions #Top_Interview_Questions #Depth_First_Search #Tree #Binary_Tree
// #Stack #Data_Structure_I_Day_10_Tree #Udemy_Tree_Stack_Queue #Big_O_Time_O(n)_Space_O(n)
// #2023_11_03_Time_456_ms_(68.42%)_Space_57.3_MB_(7.89%)

import com_github_leetcode.TreeNode
import scala.collection.mutable.ListBuffer

/*
 * Definition for a binary tree node.
 * class TreeNode(_value: Int = 0, _left: TreeNode = null, _right: TreeNode = null) {
 *   var value: Int = _value
 *   var left: TreeNode = _left
 *   var right: TreeNode = _right
 * }
 */
object Solution {
    def inorderTraversal(root: TreeNode): List[Int] = {
        if (root == null) {
            List.empty
        } else {
            val answer = ListBuffer[Int]()
            inorderTraversal(root, answer)
            answer.toList
        }
    }

    private def inorderTraversal(root: TreeNode, answer: ListBuffer[Int]): Unit = {
        if (root == null) {
            return
        }
        if (root.left != null) {
            inorderTraversal(root.left, answer)
        }
        answer += root.value
        if (root.right != null) {
            inorderTraversal(root.right, answer)
        }
    }
}
